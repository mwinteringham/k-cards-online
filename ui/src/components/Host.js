import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Button from 'react-bootstrap/Button';
import Card from './Card';
import { useGlobalState } from '../state/state';
import React, { useState, useEffect, useRef } from 'react';
import API from '../api/api';

function useInterval(callback, delay) {
  const savedCallback = useRef();

  useEffect(() => {
    savedCallback.current = callback;
  }, [callback]);

  useEffect(() => {
    function tick() {
      savedCallback.current();
    }
    if (delay !== null) {
      let id = setInterval(tick, delay);
      return () => clearInterval(id);
    }
  }, [delay]);
}

function Host(){

    const [workshopCode] = useGlobalState('workshopCode');
    const [redCards, setRedCards] = useState([]);
    const [greenCards, setGreenCards] = useState([]);
    const [yellowCards, setYellowCards] = useState([]);

    const pollForActivities = async () => {
        const res = await API.getActivity(workshopCode);

        if(res.status === 200){
            setRedCards(res.data.activity.reds);

            setGreenCards(res.data.activity.threads);

            if(res.data.activity.threads.length > 0){
                setYellowCards(res.data.activity.threads[0].subThread);
            }
        }
    }

    useEffect(() => {
        pollForActivities();
    }, [])

    useInterval(async () => {
        await pollForActivities();
    }, 2000);


    let redCardRender;
    let greenCardRender;
    let yellowCardRender;

    if(redCards.length === 0){
        redCardRender = <Col>
            <h2>No red cards found</h2>
        </Col>
    } else {
        redCardRender = redCards.map((item, i) => {
            return <Col key={i}>
                <Card type={'danger'} name={item.name} code={item.code} refresh={pollForActivities}/>
            </Col>
        });
    }

    if(greenCards.length === 0){
        greenCardRender = 
        <Row>
            <Col>
                <h2>No green cards found</h2>
            </Col>
        </Row>
    } else {
        greenCardRender = greenCards.map((item, i) => {
            if(yellowCards.length > 0){
                yellowCardRender = yellowCards.map((item, i) => {
                    return <Row key={i}>
                        <Col><Card type={'warning'} name={item.name} code={item.code} refresh={pollForActivities}/></Col>
                    </Row>
                })
            }

            return <Row key={i}>
                <Col>
                    <Card type={'success'} name={item.cardDetail.name} code={item.cardDetail.code} refresh={pollForActivities}/>
                    {i === 0 &&
                        <div>
                            {yellowCardRender}
                        </div>
                    }
                </Col>
            </Row>
        });
    }

    return (
        <div>
            <Row>
                <Col>
                    <h1>Your workshop code is: {workshopCode}</h1>
                </Col>
                <Col>
                    <Button>Close Workshop</Button>
                </Col>
            </Row>
            <Row className="mt-5">
                {redCardRender}
            </Row>
            <Row>
                <Col>
                    <h2>Activity:</h2>
                </Col>
            </Row>
            {greenCardRender}
        </div>
    )

}

export default Host;