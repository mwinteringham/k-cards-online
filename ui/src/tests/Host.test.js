import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import Host from '../components/Host';
import axios from 'axios';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';

let MockAdapter = require('axios-mock-adapter');

const mock = new MockAdapter(axios);
const history = createMemoryHistory();

test('renders the Host component', async () => {
    mock.onGet('/workshop/empty-workshop-code/activity').reply(200, {
        activity : {
            reds : [],
            threads : []
        }
    });

    const hostComponent = render(<Host />);

    await waitFor(() => screen.getAllByText(/No red cards found/i)); 

    expect(hostComponent.baseElement).toMatchSnapshot();
});

test('renders red cards', async () => {
    mock.onGet('/workshop/empty-workshop-code/activity').reply(200, {
        activity : {
            reds : [{
                code : 'abc',
                name : 'Mark'
            },{
                code : 'def',
                name : 'Mary'
            },{
                code : 'ghi',
                name : 'Sam'
            },{
                code : 'jkl',
                name : 'Jane'
            }],
            threads : []
        }
    });

    const hostComponent = render(<Host />);
    
    await waitFor(() => screen.getAllByText(/Jane/)); 

    expect(hostComponent.baseElement).toMatchSnapshot();
});

test('renders green cards', async () => {
    mock.onGet('/workshop/empty-workshop-code/activity').reply(200, {
        activity : {
            reds : [],
            threads : [{
                cardDetail : {
                    code : 'abc',
                    name : 'Green Mark'
                },
                subThread : [{
                    code : 'def',
                    name: 'Yellow Mark'
                }]
            },{
                cardDetail : {
                    code : 'ghi',
                    name : 'Green Jeff'
                }
            }]
        }
    });

    const hostComponent = render(<Host />);
    
    await waitFor(() => screen.getAllByText(/Green Jeff/)); 

    expect(hostComponent.baseElement).toMatchSnapshot();
});

test('renders yellow cards', async () => {
    mock.onGet('/workshop/empty-workshop-code/activity').reply(200, {
        activity : {
            reds : [],
            threads : [{
                cardDetail : {
                    code : 'abc',
                    name : 'Green Mark'
                },
                subThread : [{
                    code : 'def',
                    name : 'Yellow Mark'
                },{
                    code : 'ghi',
                    name : 'Yello Jeff'
                }]
            },{
                cardDetail : {
                    code : 'klm',
                    name : 'Green Jeff'
                }
            }]
        }
    });

    const hostComponent = render(<Host />);
    
    await waitFor(() => screen.getAllByText(/Green Jeff/)); 

    expect(hostComponent.baseElement).toMatchSnapshot();
});

test('Delete workshop when leaving', async () => {
    mock.onDelete('/workshop/empty-workshop-code').reply(202);

    history.push('/host');
    render(<Router history={history}><Host /></Router>);

    fireEvent.click(screen.getByTestId(/leaveWorkshop/i));

    await waitFor(() => expect(history.location.pathname).toBe('/'));
});
