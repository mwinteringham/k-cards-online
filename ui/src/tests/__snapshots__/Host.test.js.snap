// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders green cards 1`] = `
<body>
  <div>
    <div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <h1>
            Your workshop code is: 
            empty-workshop-code
          </h1>
        </div>
        <div
          class="col"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Close Workshop
          </button>
        </div>
      </div>
      <div
        class="mt-5 row"
      >
        <div
          class="col"
        >
          <h2>
            No red cards found
          </h2>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <h2>
            Activity:
          </h2>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div>
            <div
              class="card bg-success"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-body"
                >
                  <p>
                    Green Mark
                  </p>
                  <h3>
                    Card details
                  </h3>
                  <button
                    class="btn btn-primary"
                    type="button"
                  >
                    Delete
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div>
            <div
              class="row"
            >
              <div
                class="col"
              >
                <div>
                  <div
                    class="card bg-warning"
                  >
                    <div
                      class="card-body"
                    >
                      <div
                        class="card-body"
                      >
                        <p>
                          Yellow Mark
                        </p>
                        <h3>
                          Card details
                        </h3>
                        <button
                          class="btn btn-primary"
                          type="button"
                        >
                          Delete
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div>
            <div
              class="card bg-success"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-body"
                >
                  <p>
                    Green Jeff
                  </p>
                  <h3>
                    Card details
                  </h3>
                  <button
                    class="btn btn-primary"
                    type="button"
                  >
                    Delete
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`renders red cards 1`] = `
<body>
  <div>
    <div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <h1>
            Your workshop code is: 
            empty-workshop-code
          </h1>
        </div>
        <div
          class="col"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Close Workshop
          </button>
        </div>
      </div>
      <div
        class="mt-5 row"
      >
        <div
          class="col"
        >
          <div>
            <div
              class="card bg-danger"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-body"
                >
                  <p>
                    Mark
                  </p>
                  <h3>
                    Card details
                  </h3>
                  <button
                    class="btn btn-primary"
                    type="button"
                  >
                    Delete
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="col"
        >
          <div>
            <div
              class="card bg-danger"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-body"
                >
                  <p>
                    Mary
                  </p>
                  <h3>
                    Card details
                  </h3>
                  <button
                    class="btn btn-primary"
                    type="button"
                  >
                    Delete
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="col"
        >
          <div>
            <div
              class="card bg-danger"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-body"
                >
                  <p>
                    Sam
                  </p>
                  <h3>
                    Card details
                  </h3>
                  <button
                    class="btn btn-primary"
                    type="button"
                  >
                    Delete
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="col"
        >
          <div>
            <div
              class="card bg-danger"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-body"
                >
                  <p>
                    Jane
                  </p>
                  <h3>
                    Card details
                  </h3>
                  <button
                    class="btn btn-primary"
                    type="button"
                  >
                    Delete
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <h2>
            Activity:
          </h2>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <h2>
            No green cards found
          </h2>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`renders the Host component 1`] = `
<body>
  <div>
    <div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <h1>
            Your workshop code is: 
            empty-workshop-code
          </h1>
        </div>
        <div
          class="col"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Close Workshop
          </button>
        </div>
      </div>
      <div
        class="mt-5 row"
      >
        <div
          class="col"
        >
          <h2>
            No red cards found
          </h2>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <h2>
            Activity:
          </h2>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <h2>
            No green cards found
          </h2>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`renders yellow cards 1`] = `
<body>
  <div>
    <div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <h1>
            Your workshop code is: 
            empty-workshop-code
          </h1>
        </div>
        <div
          class="col"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Close Workshop
          </button>
        </div>
      </div>
      <div
        class="mt-5 row"
      >
        <div
          class="col"
        >
          <h2>
            No red cards found
          </h2>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <h2>
            Activity:
          </h2>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div>
            <div
              class="card bg-success"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-body"
                >
                  <p>
                    Green Mark
                  </p>
                  <h3>
                    Card details
                  </h3>
                  <button
                    class="btn btn-primary"
                    type="button"
                  >
                    Delete
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div>
            <div
              class="row"
            >
              <div
                class="col"
              >
                <div>
                  <div
                    class="card bg-warning"
                  >
                    <div
                      class="card-body"
                    >
                      <div
                        class="card-body"
                      >
                        <p>
                          Yellow Mark
                        </p>
                        <h3>
                          Card details
                        </h3>
                        <button
                          class="btn btn-primary"
                          type="button"
                        >
                          Delete
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="row"
            >
              <div
                class="col"
              >
                <div>
                  <div
                    class="card bg-warning"
                  >
                    <div
                      class="card-body"
                    >
                      <div
                        class="card-body"
                      >
                        <p>
                          Yello Jeff
                        </p>
                        <h3>
                          Card details
                        </h3>
                        <button
                          class="btn btn-primary"
                          type="button"
                        >
                          Delete
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div>
            <div
              class="card bg-success"
            >
              <div
                class="card-body"
              >
                <div
                  class="card-body"
                >
                  <p>
                    Green Jeff
                  </p>
                  <h3>
                    Card details
                  </h3>
                  <button
                    class="btn btn-primary"
                    type="button"
                  >
                    Delete
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
